NitroxServer/GameLogic/StoryManager.cs

```diff
@@ -0,0 +1,268 @@
+            AutoReset = true
+        };
+        ResyncTimer.Elapsed += delegate
+        {
+            playerManager.SendPacketToAllPlayers(new AuroraAndTimeUpdate(GetInitialTimeData(), false));
+        };
+    }
+
+    /// <summary>
+    /// Tells the players to start Aurora's explosion event
+    /// </summary>
+    /// <param name="countdown">Wether we should make Aurora explode instantly or after a short countdown</param>
+    public void ExplodeAurora(bool countdown)
+    {
+        // Calculations from CrashedShipExploder.OnConsoleCommand_countdownship()
+        // We add 3 seconds to the cooldown so that players have enough time to receive the packet and process it
+        AuroraCountdownTimeMs = ElapsedTimeMs + 3000;
+        AuroraWarningTimeMs = AuroraCountdownTimeMs;
+
+        if (countdown)
+        {
+            Log.Info("Aurora's explosion countdown will start in 3 seconds");
+        }
+        else
+        {
+            // Calculations from CrashedShipExploder.OnConsoleCommand_explodeship()
+            AuroraCountdownTimeMs -= 25000;
+            AuroraWarningTimeMs -= 1000;
+            Log.Info("Aurora's explosion initiated");
+        }
+
+        playerManager.SendPacketToAllPlayers(new AuroraAndTimeUpdate(GetInitialTimeData(), false));
+    }
+
+    /// <summary>
+    /// Tells the players to start Aurora's restoration event
+    /// </summary>
+    public void RestoreAurora()
+    {
+        AuroraWarningTimeMs = ElapsedTimeMs;
+        AuroraCountdownTimeMs = GenerateDeterministicAuroraTime(seed);
+
+        // We need to clear these entries from PdaLog and CompletedGoals to make sure that the client, when reconnecting, doesn't have false information
+        foreach (string eventKey in CrashedShipExploderData.AuroraEvents)
+        {
+            pdaStateData.PdaLog.RemoveAll(entry => entry.Key == eventKey);
+            storyGoalData.CompletedGoals.Remove(eventKey);
+        }
+
+        playerManager.SendPacketToAllPlayers(new AuroraAndTimeUpdate(GetInitialTimeData(), true));
+        Log.Info($"Restored Aurora, will explode again in {GetMinutesBeforeAuroraExplosion()} minutes");
+    }
+
+    /// <summary>
+    /// Calculate the future Aurora's explosion time in a deterministic manner
+    /// </summary>
+    /// <remarks>
+    /// Takes the current time into account
+    /// </remarks>
+    private double GenerateDeterministicAuroraTime(string seed)
+    {
+        // Copied from CrashedShipExploder.SetExplodeTime() and changed from seconds to ms
+        DeterministicGenerator generator = new(seed, nameof(StoryManager));
+        return elapsedTimeOutsideStopWatchMs + generator.NextDouble(2.3d, 4d) * 1200d * 1000d;
+    }
+
+    /// <summary>
+    /// Clears the already completed sunbeam events to come and broadcasts it to all players.
+    /// </summary>
+    public void StartSunbeamEvent(PlaySunbeamEvent.SunbeamEvent sunbeamEvent)
```

Comment:
Yes, this is exclusively called from the sunbeam command (redirected from the client issuing commands to manage sunbeam story) because the sunbeam events are all scheduled goals following each other.